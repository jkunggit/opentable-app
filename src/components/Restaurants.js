import React, { Component } from 'react'
import { connect } from 'react-redux'
import { FaExclamationTriangle } from 'react-icons/fa';

import { fetchRestaurants, fetchCities, changeCity, updateRefineSearch, updatePostalCode } from '../actions'
import DataListing from './DataListing'
import config from '../config'

class Restaurants extends Component {
  componentDidMount() {
    // initially load the city for the auto complete list and toronto data
    this.props.fetchCities(`${config.apiEndpoint}/cities`)
    this.props.fetchRestaurants({url: `${config.apiEndpoint}/restaurants?city=${this.props.selectedCity}`})
  }

  onChangeHandler = (e) => {
    e.persist()
    const id = e.target.id 
    const value = e.target.value
    switch(id){
      case 'changeCity':
        this.props.changeCity(value)
      break
      case 'postalCode':
        this.props.updatePostalCode(value)
      break
      case 'refineSearch':
        this.props.updateRefineSearch(value)
      break
    } 
  }  

  applySearch = (e) => {
    e.preventDefault()
    this.props.fetchRestaurants({
      url: `${config.apiEndpoint}/restaurants?city=${this.props.selectedCity}`, 
      refineSearch: this.props.refineSearch,
      postalCode: this.props.postalCode
    })
  }

  showListing = () => {
    const { restaurants, dataIsLoading } = this.props
    if(dataIsLoading) {
      return (
        <div className='loader'>Loading ...</div>
      )
    } else { 
      return restaurants.length ? 
        <DataListing restaurants={restaurants} /> 
      : 
        <div className='warning'><FaExclamationTriangle /> No restaurants matched your search criteria</div> 
    }
  }

  render () {
    const { selectedCity, refineSearch, postalCode, cities, restaurants, dataIsLoading } = this.props
   
    return (
      <main className='site-content'>
        <form>
          <div className="form-group">
            <label htmlFor='selectedCity'>City</label>
            <input 
              name='selectedCity'
              id='selectedCity'
              type='text'
              value={selectedCity} 
              onChange={this.onChangeHandler}
              list='dataListCities'
              title='City'
              placeholder='Enter City'
            />
            <datalist id="dataListCities">
              {cities.map((city, index) =>
                <option key={index} value={city} />
              )}      
            </datalist>            
          </div>
          <div className="form-group">
            <label htmlFor='refineSearch'>Address/name/area</label>
            <input 
              id='refineSearch'
              name='refineSearch'
              type='text'
              value={refineSearch} 
              onChange={this.onChangeHandler}
              title='Refine by name, address, or area'
              placeholder="Refine Search by name, address, or area"
            />          
          </div>
          <div className="form-group">
            <label htmlFor='postalCode'>Postal Code</label>
            <input 
              id='postalCode'
              name='postalCode'
              type='text'
              value={postalCode} 
              onChange={this.onChangeHandler}
              title='Search by Postal Code'
              placeholder="Search by postal Code"
            />          
          </div>         
          <div className='form-group'>
            <button id='applyBtn' onClick={this.applySearch}>Apply</button>  
          </div>     
        </form>
        {this.showListing()}
      </main>
    )
  }
}

const mapStateToProps = (state) => {
  return {
    dataIsLoading: state.dataIsLoading,
    selectedCity: state.selectedCity,
    refineSearch: state.refineSearch,
    postalCode: state.postalCode,
    cities: state.cities,
    restaurants: state.restaurants
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    changeCity: (city) => { dispatch(changeCity(city)) },
    updateRefineSearch: (val) => { dispatch(updateRefineSearch(val)) },
    updatePostalCode: (val) => { dispatch(updatePostalCode(val)) },
    fetchCities: (url) => { dispatch(fetchCities(url)) },
    fetchRestaurants: (url) => { dispatch(fetchRestaurants(url)) }
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Restaurants)
