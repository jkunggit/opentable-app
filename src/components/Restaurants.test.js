import React from 'react'
import { createStore, applyMiddleware } from 'redux'
import { Provider } from 'react-redux'
import thunk from 'redux-thunk';
import { render, cleanup, fireEvent, act, waitFor, getByText } from '@testing-library/react';
import axios from 'axios'

import reducer, { initState } from '../reducers/rootReducer'
import Restaurants from './Restaurants'


jest.mock('axios');

const renderWithRedux = (
  component,
  { initState, store = createStore(reducer, initState, applyMiddleware(thunk)) } = {}
) => {
  return {
    ...render(<Provider store={store}>{component}</Provider>),
    store,
  }
}

describe ("Restaurants", () => {
  afterEach(cleanup)

  beforeEach( () => {
    axios.get.mockResolvedValueOnce({
      data: {
        count: 3,
        cities: [
          "Toronto",
          "San Francisco",
          "New York"
        ]      
      },
    })  
    .mockResolvedValueOnce({
      data: {
        total_entries: 2,
        per_page: 25,
        current_page: 1,
        restaurants: [
          {
            id: 21307,
            name: "Scaramouche Restaurant",
            address: "1 Benvenuto Place",
            city: "Toronto",
            state: "ON",
            area: "Toronto / SW Ontario"
          },
          {
            id: 82957,
            name: "The Sultan's Tent",
            address: "49 Front St E",
            city: "Toronto",
            state: "ON",
            area: "Toronto / SW Ontario"
          },
          {
            id: 52957,
            name: "The Quick",
            address: "1345 Hummingbird Crt",
            city: "Toronto",
            state: "ON",
            area: "Toronto / SW Ontario"
          }                     
        ]     
      },
    })
    .mockResolvedValueOnce({
      data: {
        total_entries: 2,
        per_page: 25,
        current_page: 1,
        restaurants: [
          {
            id: 21307,
            name: "Scaramouche Restaurant",
            address: "1 Benvenuto Place",
            city: "Toronto",
            state: "ON",
            area: "Toronto / SW Ontario",
            postal_code: 'L1V 7T43'
          },
          {
            id: 82957,
            name: "The Sultan's Tent",
            address: "49 Front St E",
            city: "Toronto",
            state: "ON",
            area: "Toronto / SW Ontario",
            postal_code: 'A0V 7H4'
          },
          {
            id: 52957,
            name: "The Quick Pita",
            address: "1345 Hummingbird Crt",
            city: "Toronto",
            state: "ON",
            area: "Toronto / SW Ontario",
            postal_code: 'B8T 5R1'
          }                     
        ]     
      },
    })   
  })

  it ('should find all restaurants matching refine search: The Sultan\'s Tent ...', async () => {
    const { container, queryByLabelText, getByText, debug } = renderWithRedux(<Restaurants />)
    
    act(() => {
      fireEvent.change(container.querySelector('#refineSearch'), { target: { value: 'Tent'} })
    })
    
    act(() => {
      fireEvent.click(container.querySelector('#applyBtn'))
    })
      
    await waitFor( () => expect(container.querySelector('.data-listing')).not.toBe(null))
    expect(container.querySelector('.data-listing').innerHTML).toMatchSnapshot()
  })

  it ('should find all restaurants matching postal code search: The Quick Pita..', async () => {
    const { container, queryByLabelText, debug } = renderWithRedux(<Restaurants />)
    
    act(() => {
      fireEvent.change(container.querySelector('#postalCode'), { target: { value: 'b8t 5r1'} })
    })
    
    act(() => {
      fireEvent.click(container.querySelector('#applyBtn'))
    })
      
    await waitFor( () => expect(container.querySelector('.data-listing')).not.toBe(null))
    expect(container.querySelector('.data-listing').innerHTML).toMatchSnapshot()
  })

  it ('should render with redux', async () => {
    // const store = createStore(reducer, initState, applyMiddleware(thunk))
    const { container, queryByLabelText, debug } = renderWithRedux(<Restaurants />)
    
    await waitFor( () => expect(container.querySelector('.data-listing')).not.toBe(null))

    expect(queryByLabelText('City').value).toBe('Toronto')
    expect(container.querySelector('.data-listing').innerHTML).toMatchSnapshot()
  })  
})