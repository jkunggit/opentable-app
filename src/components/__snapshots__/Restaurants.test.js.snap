// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Restaurants should find all restaurants matching postal code search: The Quick Pita.. 1`] = `"<thead><tr><th>Name</th><th>Address</th><th>Area</th></tr></thead><tbody><tr><td>The Quick Pita</td><td>1345 Hummingbird Crt</td><td>Toronto / SW Ontario</td></tr></tbody>"`;

exports[`Restaurants should find all restaurants matching refine search: The Sultan's Tent ... 1`] = `"<thead><tr><th>Name</th><th>Address</th><th>Area</th></tr></thead><tbody><tr><td>The Sultan's Tent</td><td>49 Front St E</td><td>Toronto / SW Ontario</td></tr></tbody>"`;

exports[`Restaurants should render with redux 1`] = `"<thead><tr><th>Name</th><th>Address</th><th>Area</th></tr></thead><tbody><tr><td>Scaramouche Restaurant</td><td>1 Benvenuto Place</td><td>Toronto / SW Ontario</td></tr><tr><td>The Sultan's Tent</td><td>49 Front St E</td><td>Toronto / SW Ontario</td></tr><tr><td>The Quick</td><td>1345 Hummingbird Crt</td><td>Toronto / SW Ontario</td></tr></tbody>"`;
